Add missing source locations

tautschnig opened the pull request:
Linking (and also other workflows) use expression replacement. Doing so
must not destroy the source location annotated to an expression. This
left us with instructions without location.

<!---
Thank you for your contribution. Please make sure your pull request fulfils all of the below requirements. If you cannot currently tick all the boxes, but would still like to create a PR, then add the label "work in progress" and assign the PR to yourself.

If your PR fixes a bug, include the regression test(s) in the same commit as the bug fix. Else, keep commits small and orthogonal, possibly placing tests in commits of their own.
--->

- [x] Each commit message has a non-empty body, explaining why the change was made.
- n/a Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- n/a The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [x] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- n/a My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- n/a White-space or formatting changes outside the feature-related changed lines are in commits of their own.

<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

tautschnig committed:
Actually run goto-cc-multi-file regression suite

We had this folder in place for a while, but it was neither included
in the Makefile nor the CMake-based regression test execution.

tautschnig committed:
Maintain source location while replacing symbol expressions

Linking (and also other workflows) use expression replacement. Doing so
must not destroy the source location annotated to an expression. This
left us with instructions without location.

remi-delmas-3000 commented on src/util/replace_symbol.cpp:
```diff
@@ -51,7 +51,10 @@ bool replace_symbolt::replace_symbol_expr(symbol_exprt &s) const
       "\nit->second.type:\n" + it->second.type().pretty());
+  source_locationt previous_source_location{s.source_location()};
   static_cast<exprt &>(s) = it->second;
+  if(previous_source_location.is_not_nil())
```
Is it really the intuitive thing to do here ? Technically  the `symbol_exprt` is not there anymore since it's been replaced.  Maybe it could be more interesting to preserve information about where the replacement expression came from ?

remi-delmas-3000 replied:
If this  location transplant really must happen, maybe  it can be achieved by manually copying the location to the replacement expression before building the replacement map ?

remi-delmas-3000 replied:
just looking at it again, previous_source_location comes from the symbol_exprt that's being replaced

tautschnig committed:
All if-then-else goto conversion instructions have a location

We must not produce goto-program instructions without location, even
when they are considered dummy instructions: those instructions may
subsequently be used to produce a source location, which was thus
missing when converting nested if-then-else statements.

tautschnig committed:
switch-case conversion: ensure case selection has source location

The expression operand was at times lacking a source location when it
had been subject to expression simplification. Fixes the root cause in
the C front-end, and makes sure the invariant is maintained during goto
conversion.

tautschnig committed:
Function pointer removal: ensure presence of source location

Return type fix-up tried to use the source location from the
function-call code, which had just been constructed and necessarily
lacked a source location. Instead, explicitly pass a source location.

kroening commented:
I'd be curious why the various program elements end up not having a source location to begin with.

kroening commented:
It may be preferable to fix that when they are generated, as opposed to fixing it up later.

tautschnig commented:
> It may be preferable to fix that when they are generated, as opposed to fixing it up later.

Is this specifically about the commit "switch-case conversion: ensure case selection has source location" - I think all other commits do as you say, or am I missing something?

tautschnig committed:
Add exprt::with_source_location with source_locationt parameter

We already supported the fluent-style with_source_location for copying
from another expression (since 917f9a0b3d4c). Now also support passing a
source location directly.

tautschnig committed:
Reduce the number of symbol_exprt without a source location

Create or maintain source location information in symbol expressions.

tautschnig commented:
> It may be preferable to fix that when they are generated, as opposed to fixing it up later.

Thank you for insisting that this be done. I believe I have now fixed the root causes as far as feasible. (There are still places outside the language front-ends that generate symbol expressions without a source location, but we might not even have a reasonable source location to pick in those cases.)
